### ISSUE TITLE
Relay can generate queries that invalidate graphql validation

### ISSUE THREAD

**Issue by kastermester** [facebook/relay#1543]
Context: I have a schema with a union type between two different node types, let's call them `NodeA` and `NodeB`. Both `NodeA` and `NodeB` implement the relay `Node` interface, and they have a field called `state` as well, which are both enum fields, but for different enums.

When writing components that deal with either `NodeA` or `NodeB` I can obviously query the `state` field on either type. The problem arises when I then have a component which queries the union type and composes the two specific components. I end up with query like so:

```graphql
query {
  viewer {
    unionField {
      __typename
      ...F1
      ...F2
    }
  }
}

Fragment F1 on NodeA {
  id
  state
}
Fragment F2 on NodeB {
  id
  state
}
```

Now, this query does not comply with the following GraphQL validation: https://facebook.github.io/graphql/#SameResponseShape()

I have tried to get around this by aliasing the fields in the fragments spec'd on my component, but they get stripped away during compilation it seems.

**Comment by wincent** [comment]
I wonder if this could be solved by using an inline fragment spread:

```
unionField {
  ... on NodeA {
    ...F1
  }
  ... on NodeB {
    ...F2
  }
}
```

This is definitely possible in the new Relay core (not released yet, but preparations to do so are landing on the `experimental` branch), but I don't recall right now whether the old implementation elides spreads in cases like this.

**Comment by kastermester** [comment]
I tried queries of both these forms

```graphql
Fragment on UnionType {
  unionField {
    ${NodeA.getFragment('node')
    ${NodeB.getFragment('node')
  }
```
And

```graphql
Fragment on UnionType {
  unionField {
    ... on NodeA {
      ${NodeA.getFragment('node')
    }
    ... on NodeB {
      ${NodeB.getFragment('node')
    }
  }
}
```

None of these had any effect - the generated query was still invalid.

**Comment by wincent** [comment]
😞  — I was hoping that the second form would work.

We're aggressively prioritizing getting the new core out at the moment, and the new core shouldn't be afflicted by this problem. (Nobody is actually working on the old implementation at the moment, but we would review a bugfix PR if somebody wanted to provide one.)

[The spec](https://facebook.github.io/graphql/#sec-Union-type-validation) allows types in a union to implement conflicting field definitions (it doesn't *explicitly* allow it, but given that it doesn't explicitly *disallow* it, so it is therefore effectively allowed). In practical terms, this means that you you should include a `__typename` in the query on the union type to detect which one you're actually dealing with at runtime:

```
unionField {
  __typename
  ... on NodeA {
    ...F1
  }
  ... on NodeB {
    ...F2
  }
}
```

I won't close this yet, as the new core isn't out yet and somebody may wish to take a stab at a fix.

**Comment by kastermester** [comment]
Does querying the type name remove the error that the server generates?

It is ok for me to leave this be for now - I renamed my fields as that made more sense anyway. But I could see similiar issues arise around other potential fields in certain circumstances. If the new core does not have that issue then that sounds great. It just seemed to be one of the very few scenarios where a child component change could break the parent component.

I have tried to follow actively along the development of the new core, as especially the ability to do imperative code to handle mutation payloads sounds appealing to me. I am looking forward to being able to take it for a spin.
